### PR TITLE
Implementation of Conway's logic

### PR DESCRIPTION
**Pull request by lima-limon-inc**
This PR contains all the logic behing Conway's game of life. It does not contain the logic related to the GUI.

It includes the logic behind:
- [X] Changing a cells value
- [X] How many alive neighbors a cell has
- [X] Tests for both of these functions
- [ ] WIP: The logic behind calculating the new grid state.

### CODE REVIEW

File: src/conways.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+impl Grid {
+    pub fn new(size: usize) -> Grid {
+        Grid {
+            cells: vec![vec![State::Dead; size]; size],
+        }
+    }
+
+    pub fn change_state(&mut self, position: Position, state: State) {
+        let x: usize = position.0 as usize;
+        let y: usize = position.1 as usize;
```
Comment (LeanSerra) [inline src/conways.rs]: You can deconstruct the tuple `(x,y) = position` inside the function or `(x,y):Position` in the function declaration

File: src/conways.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+        let outer = &mut self.cells[y];
+        let inner = &mut outer[x];
+
+        *inner = state;
+    }
+
+    fn get_cell(&self, position: Position) -> State {
+        let x: usize = position.0 as usize;
+        let y: usize = position.1 as usize;
```
Comment (LeanSerra) [inline src/conways.rs]: Same as above

File: src/conways.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        let outer = &self.cells[y];
+        let state = outer[x];
+        return state;
+    }
+
+    pub fn update(&mut self) {
+    }
+
+    //TODO: I think this funciton's name could be improved
+    fn coordinate_from_position(&self, enumeration: u32) -> Position {
```
Comment (LeanSerra) [inline src/conways.rs]: I think that `enumeration` should have a better name

**Review by LeanSerra — APPROVED** [review]
I'll aprove it because its a WIP but I left some suggested changes

File: src/conways.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
+#[derive(Debug, Clone, PartialEq, Copy)]
+enum State {
+    Alive,
+    Dead,
+}
+
+type Position = (u32, u32);
```
Comment (dsocolobsky) [inline src/conways.rs]: I would use `usize` to avoid casting everywhere

File: src/conways.rs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            .map(|change| (change.0 + position.0 as i32, change.1 + position.1 as i32))
+            .filter(|coord| {
+                let limit: i32 = self.cells.len().try_into().unwrap();
+                coord.0 < limit && coord.1 < limit
+            });
+
+        let alive_neighbors = neighbors
+            .map(|cell| (cell.0 as u32, cell.1 as u32))
+            .filter(|cell| self.get_cell(*cell) == State::Alive)
+            .fold(0, |acc, _| acc + 1);
```
Comment (dsocolobsky) [inline src/conways.rs]: You can do `.count()` instead

File: src/conways.rs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            (1, 1),
+            (-1, 1),
+            (1, -1),
+            (-1, -1),
+        ];
+
+        let neighbors = neighbor_change
+            .iter()
+            .map(|change| (change.0 + position.0 as i32, change.1 + position.1 as i32))
+            .filter(|coord| {
```
Comment (dsocolobsky) [inline src/conways.rs]: You can destructure with  `|(x,y)|` here so it's more declarative

**Review by dsocolobsky — CHANGES_REQUESTED** [review]

**Review by damiramirez — APPROVED** [review]

**Review by LeanSerra — APPROVED** [review]
LGTM

**Pull request merged**